### PR TITLE
CDR: Run the training circuits on the simulator

### PR DESCRIPTION
**Pull request by rmlarose**
Description
-----------

For #1023 

Checklist
-----------

Check off the following once complete (or if not applicable) after opening the PR. The PR will be reviewed once this checklist is complete and all tests are passing.

- [ ] I added unit tests for new code.
- [ ] I used [type hints](https://www.python.org/dev/peps/pep-0484/) in function signatures.
- [ ] I used [Google-style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) docstrings for functions.
- [ ] I [updated the documentation](../blob/master/docs/CONTRIBUTING_DOCS.md) where relevant.

If some items remain, you can mark this a [draft pull request](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

Tips
----

- If the validation check fails:

    1. Run `make check-types` (from the root directory of the repository) and fix any [mypy](https://mypy.readthedocs.io/en/stable/) errors.

    2. Run `make check-style` and fix any [flake8](http://flake8.pycqa.org) errors.

    3. Run `make format` to format your code with the [black](https://black.readthedocs.io/en/stable/index.html) autoformatter.

  For more information, check the [Mitiq style guidelines](https://mitiq.readthedocs.io/en/stable/contributing.html#style-guidelines).
  
- Write "Fixes #XYZ" in the description if this PR fixes Issue #XYZ.

### CODE REVIEW

**Comment by rmlarose** [comment]
@L-P-B this is the fix, correct?

Tests are failing because accuracy gets worse, as expected. But the accuracy gets significantly worse in some cases. Before modifying tests I want to see if you agree on this being the fix.

File: mitiq/cdr/cdr.py [hunk 1]
```diff
@@ -158,7 +158,7 @@ def execute_with_cdr(
     results = executor.evaluate(to_run, observable)
     noisy_results = np.array(results).reshape(all_circuits_shape)
 
-    results = simulator.evaluate(all_circuits[0], observable)
+    results = simulator.evaluate(training_circuits, observable)
     ideal_results = np.array(results)
 
     # Do the regression.
```
Comment (L-P-B) [inline mitiq/cdr/cdr.py:164]: When doing the regression here it appears it is still being done on all the noisy circuits. 

This should just be done on the training circuits so it should be something like, the transpose is to ensure the shape is fitted in the linear_fit_function:

`noisy_results[1:,:].T,`

Comment (andreamari) [reply mitiq/cdr/cdr.py:164]: Makes sense to me: the fit must be based only on the training (near-clifford) results. The results associated to the circuit of interest should be used only in the last line, where the final estimate is returned.

What I don't understand is why `curve_fit` doesn't give an error if `len(noisy_results)` is not equal to `len(ideal_results)`. Anyway, this is not important I guess.

**Comment by andreamari** [comment]
Another thing that I think we should fix is the description of the `scale_factors` argument:
```
scale_factors: Factors by which to scale the noise, should not
            include 1 as this is just the original circuit. Note: When
            scale_factors is provided, the method is known as "variable-noise
            Clifford data regression."
```
I guess that, based on the code, it's actually the opposite situation. The first scale factor **must be** `1` or, at least, it is strongly recommended that the first element is `1.`

File: mitiq/cdr/cdr.py (hunk 1 above)
Comment (L-P-B) [reply mitiq/cdr/cdr.py:164]: Thats right, sorry for any confusion. 

Yes I also dont know why it does not give an error... pretty weird.

**Comment by andreamari** [comment]
Maybe failures are due to the fact that, after removing the circuit of interest from the training circuits, the CDR problem is just more difficult to solve and more training circuits are necessary (now num_training_circuits=4 in tests).  
@L-P-B, what do you think?

**Comment by L-P-B** [comment]
I would expect it to work with these circuits for even a small number of training circuits (provided there are more than 3). I think it might be something to do with the tests. Am I misunderstanding or should we always require the scalefactors to include 1. Right now in the tests it is passing a list: [3], maybe it should be [1,3] or (1,3). Sorry haven't had time to check this, just thinking outloud.

**Comment by rmlarose** [comment]
> I would expect it to work with these circuits for even a small number of training circuits (provided there are more than 3). I think it might be something to do with the tests. Am I misunderstanding or should we always require the scalefactors to include 1. Right now in the tests it is passing a list: [3], maybe it should be [1,3] or (1,3). Sorry haven't had time to check this, just thinking outloud.

It should be `scale_factors=(1, 3)` for the test with variable noise, but even the test with `scale_factors=(1,)` is failing. I'm still tinkering with this, haven't found a change that's working yet.

File: mitiq/cdr/cdr.py [hunk 2]
```diff
@@ -158,14 +159,14 @@ def execute_with_cdr(
...
     fitted_params, _ = curve_fit(
         lambda x, *params: fit_function(x, params),
-        noisy_results,
+        noisy_results[1:, :].T,
         ideal_results,
         p0=np.zeros(num_fit_parameters),
     )
-    return fit_function(noisy_results[:, 0], fitted_params)
+    return fit_function(noisy_results[0, :], fitted_params)
```
Comment (rmlarose) [inline mitiq/cdr/cdr.py:172]: ahah!

**Comment by rmlarose** [comment]
Thanks Andrea!! I fixed scale factors docstring. This looks ready to go to me.

**Review by andreamari — APPROVED** [review]

**Comment by andreamari** [comment]
@L-P-B, the changes introduced in this PR my be relevant for your simulations. Apart from `noisy_results[:, 0]` --> `noisy_results[0, :]` an further change is the handling of `num_training_params` in `execute_with_cdr()`.

**Pull request merged**